db/test/versions/0009_test.js

```diff
@@ -0,0 +1,742 @@
+const helper = require('../helper');
+const { fromNow } = require('taskcluster-client');
+const testing = require('taskcluster-lib-testing');
+const tcdb = require('taskcluster-db');
+const assert = require('assert').strict;
+const Entity = require('taskcluster-lib-entities');
+const slugid = require('slugid');
+
+const THIS_VERSION = 9;
+const PREV_VERSION = THIS_VERSION - 1;
+
+// (adapted from services/purge-cache/src/data.js)
+const CachePurgeEntity = Entity.configure({
+  version: 1,
+  partitionKey: Entity.keys.CompositeKey('provisionerId', 'workerType'),
+  rowKey: Entity.keys.StringKey('cacheName'),
+  properties: {
+    provisionerId: Entity.types.String,
+    workerType: Entity.types.String,
+    cacheName: Entity.types.String,
+    before: Entity.types.Date,
+    expires: Entity.types.Date,
+  },
+});
+
+function withEntity(db) {
+  return CachePurgeEntity.setup({
+    db,
+    serviceName: 'purge_cache',
+    tableName: 'cache_purges_entities',
+    monitor: false,
+    context: {},
+  });
+}
+
+function withDb() {
+  return tcdb.setup({
+    readDbUrl: helper.dbUrl,
+    writeDbUrl: helper.dbUrl,
+    serviceName: 'purge_cache',
+    useDbDirectory: true,
+    monitor: false,
+  });
+}
+
+function makeDocument(props = {}) {
+  return {
+    provisionerId: `${slugid.v4()}/%2F/${slugid.v4()}`,
+    workerType: `${slugid.v4()}/%2F/${slugid.v4()}`,
+    cacheName: `${slugid.v4()}/%2F/${slugid.v4()}`,
+    before: new Date(0),
+    expires: new Date(0),
+    ...props,
+  };
+}
+
+async function insertDocuments(CachePurge, num) {
+  const documents = [];
+  for (let i = 0; i < num; i++) {
+    const entry = await CachePurge.create(makeDocument({
+      provisionerId: `provisionerId-${i}`,
+      workerType: `workerType-${i}`,
+      cacheName: `cacheName-${i}`,
+      before: new Date(1),
+      expires: new Date(2),
+    }));
+
+    documents.push(entry);
+  }
+
+  return documents;
+}
+
+function compare(itemA, itemB) {
+  assert(itemA.provisionerId === itemB.provisionerId);
+  assert(itemA.workerType === itemB.workerType);
+  assert(itemA.cacheName === itemB.cacheName);
+  assert(itemA.before.toJSON() === itemB.before.toJSON());
+  assert(itemA.expires.toJSON() === itemB.expires.toJSON());
+}
+
+function runEntitiesTests(withPreviousVersion) {
+  createTests(withPreviousVersion);
+  loadTests(withPreviousVersion);
+  modifyTests(withPreviousVersion);
+  removeTests(withPreviousVersion);
+  scanTests(withPreviousVersion);
+}
+
+suite(testing.suiteName(), function() {
+  helper.withDbForVersion();
+
+  test('cache_purges table created', async function() {
+    await helper.assertNoTable('cache_purges');
+    await helper.upgradeTo(THIS_VERSION);
+    await helper.assertTable('cache_purges');
+    await helper.downgradeTo(PREV_VERSION);
+    await helper.assertNoTable('cache_purges');
+  });
+
+  test('cache purges data migrated on upgrade and downgrade', async function() {
```

Comment:
I think this can all be replaced with a call to the util function in https://github.com/djmitche/taskcluster/commit/70949f259ed237c07d9d0d5afafa31b2c1cb32bd :)